dali/sasha/saarch.cpp

```diff
@@ -85,213 +85,343 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
-                                    toXML(t,xml);
-                                    if (!cmd->addResult(xml.str()))
-                                        break;
-                                }
-                            }
                         }
                     }
-                    if (index>start+num)
+
+                    if (sortInc)
+                    {//The output should contain WUs from new to old. So, we need to reverse the order before output.
+                        Owned<CWUData> wu = new CWUData();
+                        if (outputXML || !isEmptyString(outputFormat))
+                            wu->wuTree.setown(wuTree.getClear());
+                        if (outputModifiedTime)
+                            fileIterator->getModifiedTime(wu->modifiedTime);
+                        wus.append(*wu.getClear());
+                    }
+                    else
+                    {
+                        CDateTime dt;
+                        if (outputModifiedTime)
+                            fileIterator->getModifiedTime(dt);
+
+                        if (!setOutput(cmd, outputFields, wuTree, dt))
+                        {
+                            overflowed = true;
+                            break;
+                        }
+                    }
+                    wuCount++;
+                    if (wuCount == numWUs)
                         break;
                 }
+                if (overflowed || (wuCount == numWUs))
+                    break;
             }
+            if (sortInc)
+            {
+                ForEachItemInRev(i, wus)
+                {
+                    CWUData &wuData = wus.item(i);
+                    if (!setOutput(cmd, outputFields, wuData.wuTree, wuData.modifiedTime))  //Log an error?
+                        break;
+                }
+            }
+
+            if (hasWUSOutput)
+                cmd->setWUSresult(WUSbuf);
         }
-    }
-    if (cmd->getOnline()) {
-        if (haswusoutput)
+    };
+
+    Owned<CWUReader> reader = new CWUReader(cmd, mask);
+    if (!cmd->getOnline())
+        reader->getWUs(cmd, cmd->queryBeforeWU(), cmd->queryAfterWU(), cmd->getLimit());
+    else
+    {//The following code is from the pre-existing code.
+        bool dfu = cmd->getDFU();
+        const char *beforedt = cmd->queryBefore();
+        const char *afterdt = cmd->queryAfter();
+        const char *owner = cmd->queryOwner();
+        const char *state = cmd->queryState();
+        const char *cluster = cmd->queryCluster();
+        const char *jobname = cmd->queryJobName();
+        const char *outputformat = cmd->queryOutputFormat();
+        unsigned start = cmd->getStart(); 
+        unsigned num = cmd->getLimit();
+        StringBuffer before;
+        StringBuffer after;
+        StringBuffer tmppath;
+        mkDateCompare(dfu,afterdt,after,'0');
+        mkDateCompare(dfu,beforedt,before,'9');
```

Comment:
>//The following code is from the pre-existing code.

Any reason not to move the bit that isn't exactly the same, into CWUReader, as a getOnlineWus() or similar?
The code in WUReader does all the above and some of the below and your new code is cleanedup
Should avoid this duplication, it will cause maintenance problems having 2 copies later.

Reply:
I was assuming that the code to get online WUs here should no longer be used.  I guess I should not assume that. I will move the code into a new getOnlineWus().

Reply:
@wangkx - am not sure what you mean..
Why would this code not be called? What would prevent it? Perhaps I am missing something somewhere that means it would not be called ?

Reply:
Sorry for confusing you. My previous comment just tried to explain that why I did not move the getOnlineWus() code into CWUReader and I will move it now.